<?xml version="1.0"?>
<launch> 

  <!-- <arg name="bagdir" default="/home/pmvanderburg/noetic-husky/bop_ros_ws/src/Husky_scripts/bagfiles/" /> -->
  <!-- <arg name="bagdir" default="/home/pmvanderburg/noetic-husky/bop_ros_ws/src/Husky_scripts/bagfiles/20220705" /> -->
  <!-- <arg name="bagdir" default="/media/pmvanderburg/T71/bagfiles/20220705" /> -->
  <arg name="date" default="20220705" />
  <arg name="sequence_number" default="000000" />
  <arg name="sync_delay" default="0" />
  <!-- <arg name="suffix" default="synced_delay_$(arg sync_delay)" /> -->
  <arg name="suffix" default="edit_delay_$(arg sync_delay)_nosync" />
  <arg name="bagdir" default="/media/pmvanderburg/T7/bagfiles/20220705" />
  <arg name="bagpath" default="$(arg bagdir)/$(arg date)_exp_$(arg sequence_number)_$(arg suffix).bag" />
  <arg name="dummynr" default="" />
  <arg name="dummybagpath" default="$(arg bagdir)/$(arg date)_exp_$(arg sequence_number)_dummy-$(arg dummynr).bag" />
  <arg name="dummy_crop" default="1" />

  <!-- bop_datasets path-->
  <!-- <arg name="dataset_path" default="/home/pmvanderburg/noetic-husky/datasets/bop_datasets/husky" />-->
  <arg name="dataset_path" default="/media/pmvanderburg/T7/bop_datasets/husky" />

  <arg name="bagrate" default="1" />
  <arg name="json_frequency" default="500" /> <!-- 50hz works well for 20220705_exp_0001**_synced  -->
  <arg name="zed_focal_len" default="0.00212" />
  <arg name="cam_pitch" default="24.2" />  <!-- this value works well for 20220705_exp_00004*_synced  -->
  <!-- <arg name="sleepbf" default="3.95" />  this value works well for 20220705_exp_000001_synced  -->
  <arg name="sleepbf" default="1.2" /> <!-- this value works well for 20220705_exp_0001**_synced  -->
  <!-- <arg name="bagdelay" default="4.5" /> this value works well for 20220705_exp_000001_synced  -->
  <!-- <arg name="bagdelay" default="1.9" />  this value works well for 20220705_exp_0001**_synced  -->
  <arg name="bagdelay" default="2.0" /> <!-- this value works well for 20220705_exp_0001**_synced  -->
  <arg name="DEBUG" default="0" />
  <arg name="node_start_delay" default="2.5" />  
  <arg name="split_type" default="train_tmp" />  

  <!-- convert to params for use across multiple nodes-->
  <param name="bagpath" type="string" value="$(arg bagpath)" />
  <param name="dataset_path" type="" value="$(arg dataset_path)" />
  <param name="sequence_number" type="" value="$(arg sequence_number)" />
  <param name="split_type" type="" value="$(arg split_type)" />
  <param name="sync_delay" type="" value="$(arg sync_delay)" />

    <!-- create dummybag for consistent tf_lookup behaviour -->
<!--    <node pkg="bop_ros_conversion" type="rewrite_header_node.py" name="ros_rewrite_header_delay" respawn="false" output="screen" >
    -->
  <!--<param name="src" type="string" value="$(arg bagpath)" /> -->
<!--      <param name="tgt" type="string" value="$(arg dummybagpath)" />
  <param name="crop" type="bool"  value="$(arg dummy_crop)" />
    </node>
    -->

    <!-- bag_to_images -->
    <node pkg="bop_ros_conversion" type="bag_to_image_node.py" name="rosbag_to_images" respawn="false" output="screen" >
      <param name="lefttop" type="string" value="/sync/left/image_rect" />
      <param name="leftcamtop" type="string" value="/sync/left/camera_info" />
      <param name="righttop" type="string" value="/sync/right/image_rect" />
      <param name="rightcamtop" type="string" value="/sync/right/camera_info" />
    </node>

    <!-- from optitrack odom to tf2 broadcasters for husky and box -->
    <node name="husky_tf2_broadcaster" pkg="bop_ros_conversion" type="odom_to_tf2_broadcaster.py" respawn="false" output="screen" >
      <param name="agentname1" type="string" value="Husky" /> <!-- name for detecting it in the sync topic -->
      <param name="agentname2" type="string" value="robot" /> <!-- child frame id for the tf topic -->
    </node>
  
    <node name="box_tf2_broadcaster" pkg="bop_ros_conversion" type="odom_to_tf2_broadcaster.py" respawn="false" output="screen" >
      <param name="agentname1" type="string" value="MMbox" /> <!-- name for detecting it in the sync topic -->
      <param name="agentname2" type="string" value="object" /> <!-- child frame id for the tf topic --> 
    </node>
    
    <!-- static transforms below to publish the different coordinate frames (husky base_plate and zed sensors and box centre) 
    <node pkg="tf" type="static_transform_publisher" name="husky_opti_to_base_static_broadcaster" args="-.013881934 .009032268 -.10728907 0 0 0 1 Husky_opti_link Husky_base_link $(arg static_ms)" />
    <node pkg="tf" type="static_transform_publisher" name="husky_base_to_zed_right_static_broadcaster" args="-0.1593366 -.1050237 0.3558767  0 0 0 1 Husky_base_link Husky_zed_right2 $(arg static_ms)" />
    <node pkg="tf" type="static_transform_publisher" name="husky_base_to_zed_left_static_broadcaster" args="-0.1553935 .013355832 .35680165  0 0 0 1 Husky_base_link Husky_zed_left2 $(arg static_ms)" />
    <node pkg="tf" type="static_transform_publisher" name="box_opti_to_top_ctr_static_broadcaster" args="-0.0091251 0.003978 -0.001673 0 0 0 1 MMbox_opti_link MMbox_top_ctr_2 $(arg static_ms)" />
    <node pkg="tf" type="static_transform_publisher" name="box_top_ctr_to_ctr_static_broadcaster" args="0 0 -0.182 0 0 0 1 MMbox_top_ctr_2 MMbox_base_link $(arg static_ms)" />
    -->

    <node name="static_tf2_broadcaster_base" pkg="bop_ros_conversion" type="odom_to_static_broadcaster.py" respawn="false" output="screen" >
      <param name="topicname" type="string" value="/sync/Husky/pose" />          <!-- name for detecting it in the sync topic -->
      <param name="parentname" type="string" value="Husky_opti_link" />          <!-- child frame id for the tf topic --> 
      <param name="childname" type="string" value="Husky_base_link" />           <!-- name for detecting it in the sync topic -->
    </node>
    <node name="static_tf2_broadcaster_zedl" pkg="bop_ros_conversion" type="odom_to_static_broadcaster.py" respawn="false" output="screen" >
      <param name="topicname"  type="string" value="/sync/Husky/pose" />       <!-- name for detecting it in the sync topic -->
      <param name="parentname" type="string" value="Husky_base_link" />        <!-- parent frame id for the tf topic -->
      <param name="childname"  type="string" value="Husky_zed_left_mk_link" /> <!-- child frame id for the tf topic -->
      <param name="campitch"                 value="$(arg cam_pitch)" />       <!-- pitching angle of the zed camera -->
    </node>
    <node name="static_tf2_broadcaster_zedsl" pkg="bop_ros_conversion" type="odom_to_static_broadcaster.py" respawn="false" output="screen" >
      <param name="topicname"  type="string" value="/sync/Husky/pose" />           <!-- name for detecting it in the sync topic -->
      <param name="parentname" type="string" value="Husky_zed_left_mk_link" />     <!-- parent frame id for the tf topic -->
      <param name="childname"  type="string" value="Husky_zed_left_sensor_link" /> <!-- child frame id for the tf topic -->
      <param name="focal_length"             value="$(arg zed_focal_len)" />       <!-- focal length of the zed camera -->
    </node>
    <node name="static_tf2_broadcaster_zedr" pkg="bop_ros_conversion" type="odom_to_static_broadcaster.py" respawn="false" output="screen" >
      <param name="topicname"  type="string" value="/sync/Husky/pose" />           <!-- name for detecting it in the sync topic -->
      <param name="parentname" type="string" value="Husky_zed_left_mk_link" />          <!-- parent frame id for the tf topic -->
      <param name="childname"  type="string" value="Husky_zed_right_mk_link" /> <!-- child frame id for the tf topic -->
    </node>
    <node name="static_tf2_broadcaster_zedsr" pkg="bop_ros_conversion" type="odom_to_static_broadcaster.py" respawn="false" output="screen" >
      <param name="topicname"  type="string" value="/sync/Husky/pose" />            <!-- name for detecting it in the sync topic -->
      <param name="parentname" type="string" value="Husky_zed_left_sensor_link" />  <!-- parent frame id for the tf topic -->
      <param name="childname"  type="string" value="Husky_zed_right_sensor_link" /> <!-- child frame id for the tf topic -->
      <param name="focal_length" value="$(arg zed_focal_len)" />                    <!-- focal length of the zed camera -->
      <param name="campitch" value="$(arg cam_pitch)" />                            <!-- pitching angle of the zed camera -->
    </node>
    <node name="static_tf2_broadcaster_boxctr" pkg="bop_ros_conversion" type="odom_to_static_broadcaster.py" respawn="false" output="screen" >
      <param name="topicname"  type="string" value="/sync/MMbox/pose" />           <!-- name for detecting it in the sync topic -->
      <param name="parentname" type="string" value="MMbox_opti_link" />          <!-- parent frame id for the tf topic -->
      <param name="childname"  type="string" value="MMbox_base_link" /> <!-- child frame id for the tf topic -->
    </node>

    <node name="odom_to_json_listener" pkg="bop_ros_conversion" type="odom_to_json_listener_nosync.py" launch-prefix="bash -c 'sleep $(arg node_start_delay); $0 $@' " respawn="false" output="screen" >
      <param name="parent_frame" type= "string" value="Husky_zed_left_sensor_link"/>
      <param name="child_frame" type= "string" value="MMbox_base_link"/>
      <param name="lookup_rate" value="$(arg json_frequency)"/>
      <param name="sleeptime" value="$(arg sleepbf)"/>
      <param name="debug" value="$(arg DEBUG)"/>
    </node>

    <node pkg="rosbag" type="play" name="rosbag" args="-r $(arg bagrate) -d $(arg bagdelay) $(arg bagpath)" launch-prefix="bash -c 'sleep $(arg node_start_delay); $0 $@' " />

</launch>
